// fancy-progress/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var progress_exports = {};
__export(progress_exports, {
  SingleBar: () => SingleBar,
  create: () => create,
  render: () => import_ink2.render
});
module.exports = __toCommonJS(progress_exports);
var import_react = __toESM(require("react"));
var import_ink = require("ink");
var import_ink2 = require("ink");
function SingleBar(props) {
  const width = 30;
  const fill = Math.floor(width * props.progress);
  return /* @__PURE__ */ import_react.default.createElement(import_ink.Text, null, /* @__PURE__ */ import_react.default.createElement(import_ink.Newline, null), /* @__PURE__ */ import_react.default.createElement(import_ink.Text, { color: props.color }, "\u25CF ", props.name, " "), /* @__PURE__ */ import_react.default.createElement(import_ink.Text, { backgroundColor: props.color }, " ".repeat(Math.min(fill, width))), /* @__PURE__ */ import_react.default.createElement(import_ink.Text, { backgroundColor: "white" }, " ".repeat(Math.max(width - fill, 0))), /* @__PURE__ */ import_react.default.createElement(import_ink.Text, { color: "white" }, " ", Math.floor(props.progress * 100), "%"), /* @__PURE__ */ import_react.default.createElement(import_ink.Newline, null), /* @__PURE__ */ import_react.default.createElement(import_ink.Text, { color: "grey" }, "  ", props.description));
}
;
SingleBar.defaultProps = {
  color: "green",
  name: "progress",
  progress: 0,
  description: ""
};
const currentBars = [];
let inited = false;
let update = (bars) => {
};
function Manager() {
  const [bars, setBars] = (0, import_react.useState)("[]");
  update = setBars;
  return /* @__PURE__ */ import_react.default.createElement(import_ink.Box, { flexDirection: "column" }, JSON.parse(bars).map((bar, i) => {
    return /* @__PURE__ */ import_react.default.createElement(SingleBar, { key: i, ...bar });
  }));
}
function create(name, color) {
  if (!inited) {
    (0, import_ink.render)(/* @__PURE__ */ import_react.default.createElement(Manager, null));
    inited = true;
  }
  const bar = {
    name,
    color,
    progress: 0,
    description: ""
  };
  currentBars.push(bar);
  return {
    update(progress, description) {
      bar.progress = progress;
      bar.description = description;
      update(JSON.stringify(currentBars));
    },
    done() {
      currentBars.splice(currentBars.indexOf(bar), 1);
      update(JSON.stringify(currentBars));
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  SingleBar,
  create,
  render
});
